### PR TITLE
[Storage] FOR compression

### PR DESCRIPTION
**Pull request by Tishj**
# Frame Of Reference Compression

This PR adds FOR (Frame-Of-Reference) compression to the currently implemented Bitpacking compression.

Frame of reference compression works by offsetting all values by a common value (the frame of reference).
To select the value to be used as Frame of Reference, (usually) the lowest value is used.

This resulting value in general will have a better bitpacking compression ratio than the original.
(imagine how [2000, 2005, 2003] can be changed to [0,5,3] (with 2000 as FOR) which can be compressed to just 3 bits per value.

## Compression ratio formula comparison:
(Lets use 8 bits for simplicity here)

Bitpacking: total bits = `8 + (n * ((0|1) + bit_width))`
8 bits to store the bit_width number
0|1 bits to represent the sign (0 if the value type is unsigned - we don't need to store sign information)
bit_width - the amount of bits we need to represent the highest number

Frame-of-reference: total bits = `8 + 8 + (n * ((0|1) + bit_width))`
8 bits to store the bit_width number
8 bits to store the frame_of_reference
0|1 bits to represent the sign (0 if the value type is unsigned - we don't need to store sign information)
bit_width - the amount of bits we need to represent the highest number

### Signed Overflow

The main problem we've faced is overflow:
(8 bit values are used here for ease of writing)
Imagine the example: `[-128, 5]`.
`-128` is our lowest value, and offsetting all values by that value results in: `[0, 133]` - which overflows 🔴 

### Reject the compression method

When subtracting frame of reference from a value would cause overflow, that means that we need at least all the bits available to us, to represent this difference - In which case bitpacking would be worse than leaving the data uncompressed.

So in these scenarios we bounce the bitpacking compression ratio from the available compression options.

As a result nothing really changes about bitpacking, other than adding an additional step to compression and decompression, and an additional constraint that was kind of already there (if `bit_width == sizeof(T)` we would never use bitpacking), but is now a harder constraint.

## Additional changes

- Throwing a FatalException will now invalidate the database, resulting in the failure of every subsequent query.
- InternalException now inherits from FatalException
- `force_compression` pragma now falls back to UNCOMPRESSED if the forced method could not be used

### CODE REVIEW

**Comment by Tishj** [comment]
Other options considered:

Treating signed values as unsigned, so no overflow could occur when subtracting the frame-of-reference.
As a result of this, we would lose sign compression.
Meaning that [-20, 3, 5] would compress much worse than before:
```
NEW: [-20, 3, 5] -> unsigned [236, 3, 5] -> FOR [233, 0, 5]
OLD: [-20, 3, 5] -> sign compression [20,3,5]
```

An attempt to counter-act this issue is to use zig-zag encoding (only for signed numbers), which essentially doubles every positive value, and places every positive numbers negative counterpart in the gaps
```
0 -> 0,
-1 -> 1,
1 -> 2,
-2 -> 3,
2 -> 4
```

This could also have worked, as this would allow us to get rid of the extra bit we need to store the original sign of the number.
But since every number is doubled, this would introduce this extra bit again anyways, so effectively this would not change anything.

File: src/common/exception.cpp [hunk 1]
```diff
@@ -213,10 +213,10 @@ SequenceException::SequenceException(const string &msg) : Exception(ExceptionTyp
 InterruptException::InterruptException() : Exception(ExceptionType::INTERRUPT, "Interrupted!") {
 }
 
-FatalException::FatalException(const string &msg) : Exception(ExceptionType::FATAL, msg) {
+FatalException::FatalException(ExceptionType type, const string &msg) : Exception(type, msg) {
 }
 
-InternalException::InternalException(const string &msg) : Exception(ExceptionType::INTERNAL, msg) {
+InternalException::InternalException(const string &msg) : FatalException(ExceptionType::INTERNAL, msg) {
```
Comment (samansmink) [inline src/common/exception.cpp:219]: I don't really understand the reason for this change, why is this needed?

File: src/include/duckdb/common/bitpacking.hpp [hunk 2]
```diff
@@ -178,7 +178,31 @@ class BitpackingPrimitives {
 	}
 
 	template <class T>
-	static void UnPackGroup(data_ptr_t dst, data_ptr_t src, bitpacking_width_t width,
+	static void ApplyFrameOfReference(data_ptr_t dst, T frame_of_reference) {
+		if (!frame_of_reference) {
+			return;
+		}
+		for (idx_t i = 0; i < BITPACKING_ALGORITHM_GROUP_SIZE; i++) {
+			T value = Load<T>(dst + (i * sizeof(T)));
```
Comment (samansmink) [inline src/include/duckdb/common/bitpacking.hpp:186]: Do we need Load and Store here? I think the dst here should always be aligned, no?

File: src/storage/compression/bitpacking.cpp [hunk 3]
```diff
@@ -179,12 +222,14 @@ struct BitpackingCompressState : public CompressionState {
...
 	}
 
-	void WriteValues(T *values, bitpacking_width_t width, idx_t count) {
+	void WriteValues(T *values, bitpacking_width_t width, T frame_of_reference, idx_t count) {
 		// TODO we can optimize this by stopping early if count < BITPACKING_WIDTH_GROUP_SIZE
 		BitpackingPrimitives::PackBuffer<T, false>(data_ptr, values, count, width);
 		data_ptr += (BITPACKING_WIDTH_GROUP_SIZE * width) / 8;
 
 		Store<bitpacking_width_t>(width, width_ptr);
+		width_ptr -= sizeof(T);
```
Comment (samansmink) [inline src/storage/compression/bitpacking.cpp:231]: width_ptr should be renamed probably as it no longer points to just widths, but to width + for.

File: src/include/duckdb/common/bitpacking.hpp [hunk 4]
```diff
@@ -178,7 +178,31 @@ class BitpackingPrimitives {
...
+		for (idx_t i = 0; i < BitpackingPrimitives::BITPACKING_ALGORITHM_GROUP_SIZE; ++i) {
+			T value = Load<T>(dst + i * sizeof(T));
+			value = value & ((((T_U)1) << width) - ((T_U)1));
+			T result = (value ^ mask) - mask;
+			Store(result, dst + i * sizeof(T));
+		}
+	}
+
+	template <class T>
+	static void UnPackGroup(data_ptr_t dst, data_ptr_t src, bitpacking_width_t width, T frame_of_reference,
```
Comment (samansmink) [inline src/include/duckdb/common/bitpacking.hpp:205]: This function is also used in places where we don't currently do FOR i think, it would be nice to not have the overhead of applying 0 there

**Review by samansmink — CHANGES_REQUESTED** [review]
Hey @Tishj, looks very good! I think a few small things can be improved. I added a few comments, some of which we already discussed a bit this morning!

File: src/common/exception.cpp (hunk 1 above)
Comment (Tishj) [reply src/common/exception.cpp:219]: If I'm not mistaken this came out of a discussion I had with Mytherin, I think the general idea is that if an Internal Exception is triggered, that's similar to an assertion. At which point we cant guarantee that the database is in a valid state anymore.

File: src/include/duckdb/common/bitpacking.hpp (hunk 2 above)
Comment (Tishj) [reply src/include/duckdb/common/bitpacking.hpp]: Yep you're right, changed it

File: src/include/duckdb/common/bitpacking.hpp (hunk 4 above)
Comment (Tishj) [reply src/include/duckdb/common/bitpacking.hpp]: I was experiencing a bug caused by addition on random memory (because of the algorithm group size) so I moved the applying of the frame of reference out of here, which should also fix this issue

File: src/storage/compression/bitpacking.cpp (hunk 3 above)
Comment (Tishj) [reply src/storage/compression/bitpacking.cpp:253]: would `metadata_ptr` work?

Comment (samansmink) [reply src/storage/compression/bitpacking.cpp:253]: sounds good to me!

File: test/sql/storage/compression/bitpacking/bitpacking_fatal_forced.test_coverage [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+load __TEST_DIR__/test_bitpacking.db
+
+statement ok
+PRAGMA force_compression = 'bitpacking'
+
+statement ok
+CREATE TABLE test (a INT64, b INT32, c INT16, d INT8);
+
+statement ok
+INSERT INTO test SELECT i, i, i, i FROM range(-128, 128) tbl(i);
```
[the remark this answers is not in the document]
Comment (Tishj) [reply test/sql/storage/compression/bitpacking/bitpacking_fatal_forced.test_coverage]: I have made a change to `force_compression`'s behavior so it falls back to UNCOMPRESSED instead of failing altogether.

As a result of this I am checking the compression used with `storage_info` for each column separately to confirm the correct compression method is used.

Would you still like me to make more tests for this or does that meet the requirements of your request?

[the remark this answers is not in the document]
Comment (samansmink) [reply test/sql/storage/compression/bitpacking/bitpacking_fatal_forced.test_coverage]: The tests you added there lgtm!

File: src/storage/compression/bitpacking.cpp [hunk 6]
```diff
@@ -20,14 +21,47 @@ static constexpr const idx_t BITPACKING_WIDTH_GROUP_SIZE = 1024;
...
 struct EmptyBitpackingWriter {
 	template <class T>
-	static void Operation(T *values, bool *validity, bitpacking_width_t width, idx_t count, void *data_ptr) {
+	static void Operation(T *values, bool *validity, bitpacking_width_t width, T frame_of_reference, idx_t count,
+	                      void *data_ptr) {
+	}
+};
+
+template <class T>
+struct BufferMinMax {
```
Comment (samansmink) [inline src/storage/compression/bitpacking.cpp:30]: I think this does not need to be a separate Struct. 

I think this would be simplified by removeing the `BufferMinMax` struct by having the bitpacking state contain a `T min`, `T max` and `bool min_max_set` property.  Then a `TryUpdateMinMax(T value)` combined with a `ResetMinMax()` that is called from the Flush.

Also it seems a bit cleaner to first do TrySubtractOperator, and then updating the min/max values when this succeeds because now a `TryUpdate` that fails, still sets the failing min/max values.

File: src/storage/compression/bitpacking.cpp [hunk 7]
```diff
@@ -36,21 +70,36 @@ struct BitpackingState {
...
+	}
+
 	template <class OP>
 	void Flush() {
-		bitpacking_width_t width = BitpackingPrimitives::MinimumBitWidth<T>(compression_buffer, compression_buffer_idx);
-		OP::Operation(compression_buffer, compression_buffer_validity, width, compression_buffer_idx, data_ptr);
-		total_size += (BITPACKING_WIDTH_GROUP_SIZE * width) / 8 + sizeof(bitpacking_width_t);
+		T frame_of_reference = compression_buffer_minmax.Minimum();
+		SubtractFrameOfReference(frame_of_reference);
+		bitpacking_width_t width = BitpackingPrimitives::MinimumBitWidth(compression_buffer, compression_buffer_idx);
```
Comment (samansmink) [inline src/storage/compression/bitpacking.cpp:86]: Can we gain an extra bit of compression here using the fact that even for signed values we can no longer have negative values here?

File: src/storage/compression/bitpacking.cpp [hunk 8]
```diff
@@ -263,11 +323,16 @@ struct BitpackingScanState : public SegmentScanState {
 	data_ptr_t current_width_group_ptr;
 	data_ptr_t bitpacking_width_ptr;
 	bitpacking_width_t current_width;
+	T current_frame_of_reference;
 
 public:
+	//! Loads the current group header, and sets pointer to next header
```
Comment (samansmink) [inline src/storage/compression/bitpacking.cpp:329]: `LoadCurrentMetaData`, since we load both width and for

File: src/storage/compression/bitpacking.cpp [hunk 9]
```diff
@@ -316,28 +390,24 @@ void BitpackingScanPartial(ColumnSegment &segment, ColumnScanState &state, idx_t
...
-			scan_state.bitpacking_width_ptr -= sizeof(bitpacking_width_t);
 			scan_state.LoadCurrentBitWidth();
 			return;
 		}
 	}
 
-	// Determine if we can skip sign extension during compression
-	auto &nstats = (NumericStatistics &)*segment.stats.statistics;
-	bool skip_sign_extend = std::is_signed<T>::value && nstats.min >= 0;
+	bool skip_sign_extend = true;
```
Comment (samansmink) [inline src/storage/compression/bitpacking.cpp:403]: can we add a comment here why we can always skip sign extending?

**Review by samansmink — CHANGES_REQUESTED** [review]

File: src/storage/compression/bitpacking.cpp (hunk 7 above)
Comment (Tishj) [reply src/storage/compression/bitpacking.cpp:89]: Ah yea mean by providing the templated bool to MinimumBitWidth that we talked about?
Yes I'll see if I can get that to work

Comment (Tishj) [reply src/storage/compression/bitpacking.cpp:89]: We can optimize even further, as we don't need to do the additional loop that FindMinimumBitWidth is doing to get the minimum/maximum value, because we already keep track of that

Comment (Tishj) [reply src/storage/compression/bitpacking.cpp]: Hmm I don't know if I made a mistake, or this is again due to counting blocks instead of bytes.

The new theoretical compression ratio for `(2000 + i%2) for i in range(1024)` is 56.88
We're getting 41.0 now

Comment (Tishj) [reply src/storage/compression/bitpacking.cpp]: Thanks for explaining it to me, I think I understand the relation between row groups, vectors, blocks and segments a lot better now!

File: src/storage/table/column_data_checkpointer.cpp [hunk 10]
```diff
@@ -117,19 +124,25 @@ unique_ptr<AnalyzeState> ColumnDataCheckpointer::DetectBestCompressionMethod(idx
 
 	// now that we have passed over all the data, we need to figure out the best method
 	// we do this using the final_analyze method
-	unique_ptr<AnalyzeState> state;
+	unique_ptr<AnalyzeState> state = nullptr;
```
Comment (Mytherin) [inline src/storage/table/column_data_checkpointer.cpp:127]: nit: `unique_ptr` is always initialized with `nullptr`

**Review by Mytherin — APPROVED** [review]
Thanks for the updates, LGTM. @samansmink any remaining comments?

**Comment by samansmink** [comment]
The PR LGTM! would be interesting to see the benchmark results compared with master for:
```
benchmark/micro/compression/bitpacking/bitpacking_read.benchmark
benchmark/micro/compression/bitpacking/bitpacking_write.benchmark
```

**Comment by Mytherin** [comment]
Thanks!
